From 7c678b145196c30cd00e26b91f4e53ae2eab1122 Mon Sep 17 00:00:00 2001
From: Kreijstal <elektrischrainbow@gmail.com>
Date: Wed, 12 Aug 2026 18:00:00 +0200
Subject: [PATCH] Fix MinGW FFI exports

GCC ignores dllexport when it appears before the extern "C" linkage
specification. Put LEAN_EXPORT on the declaration so the split Lean DLLs can
import these functions from libleanshared_1.
---
 src/util/ffi.cpp        | 8 ++++----
 stage0/src/util/ffi.cpp | 8 ++++----
 2 files changed, 8 insertions(+), 8 deletions(-)

diff --git a/src/util/ffi.cpp b/src/util/ffi.cpp
index 8adb610ef..8f0b70dc3 100644
--- a/src/util/ffi.cpp
+++ b/src/util/ffi.cpp
@@ -9,19 +9,19 @@ Author: Sebastian Ullrich
 #include "runtime/string_ref.h"

 namespace lean {
-LEAN_EXPORT extern "C" object * lean_get_leanc_extra_flags(object *) {
+extern "C" LEAN_EXPORT object * lean_get_leanc_extra_flags(object *) {
     return lean_mk_string("@LEANC_EXTRA_CC_FLAGS@");
 }

-LEAN_EXPORT extern "C" object * lean_get_leanc_internal_flags(object *) {
+extern "C" LEAN_EXPORT object * lean_get_leanc_internal_flags(object *) {
     return lean_mk_string("@LEANC_INTERNAL_FLAGS@");
 }

-LEAN_EXPORT extern "C" object * lean_get_linker_flags(uint8 link_static) {
+extern "C" LEAN_EXPORT object * lean_get_linker_flags(uint8 link_static) {
     return lean_mk_string(link_static ? "@LEANC_STATIC_LINKER_FLAGS@ @LEAN_EXTRA_LINKER_FLAGS@" : "@LEANC_SHARED_LINKER_FLAGS@ @LEAN_EXTRA_LINKER_FLAGS@");
 }

-LEAN_EXPORT extern "C" object * lean_get_internal_linker_flags(object *) {
+extern "C" LEAN_EXPORT object * lean_get_internal_linker_flags(object *) {
     return lean_mk_string("@LEANC_INTERNAL_LINKER_FLAGS@");
 }
 }
diff --git a/stage0/src/util/ffi.cpp b/stage0/src/util/ffi.cpp
index 8adb610ef..8f0b70dc3 100644
--- a/stage0/src/util/ffi.cpp
+++ b/stage0/src/util/ffi.cpp
@@ -9,19 +9,19 @@ Author: Sebastian Ullrich
 #include "runtime/string_ref.h"

 namespace lean {
-LEAN_EXPORT extern "C" object * lean_get_leanc_extra_flags(object *) {
+extern "C" LEAN_EXPORT object * lean_get_leanc_extra_flags(object *) {
     return lean_mk_string("@LEANC_EXTRA_CC_FLAGS@");
 }

-LEAN_EXPORT extern "C" object * lean_get_leanc_internal_flags(object *) {
+extern "C" LEAN_EXPORT object * lean_get_leanc_internal_flags(object *) {
     return lean_mk_string("@LEANC_INTERNAL_FLAGS@");
 }

-LEAN_EXPORT extern "C" object * lean_get_linker_flags(uint8 link_static) {
+extern "C" LEAN_EXPORT object * lean_get_linker_flags(uint8 link_static) {
     return lean_mk_string(link_static ? "@LEANC_STATIC_LINKER_FLAGS@ @LEAN_EXTRA_LINKER_FLAGS@" : "@LEANC_SHARED_LINKER_FLAGS@ @LEAN_EXTRA_LINKER_FLAGS@");
 }

-LEAN_EXPORT extern "C" object * lean_get_internal_linker_flags(object *) {
+extern "C" LEAN_EXPORT object * lean_get_internal_linker_flags(object *) {
     return lean_mk_string("@LEANC_INTERNAL_LINKER_FLAGS@");
 }
 }
--
2.51.0
